lab_segmentation.py

```diff
@@ -0,0 +1,191 @@
+    samples = extract_training_samples(feature_image, sampling_rectangle)
+    model = model_type(samples)
+
+    # Set up a simple gui for the lab (based on OpenCV highgui) and run the main loop.
+    with SegmentationLabGui(initial_thresh_val, max_distance) as gui:
+        while True:
+            # Read next frame.
+            success, frame = cap.read()
+            if not success:
+                break
+
+            # Extract features.
+            feature_image = extract_features(frame)
+
+            # Update if using adaptive model
+            if use_adaptive_model:
+                new_samples = extract_training_samples(feature_image, sampling_rectangle)
+                update_samples(samples, new_samples, adaptive_update_ratio)
+                model = model_type(samples)
+
+            # Compute how well the pixel features fit with the model.
+            mahalanobis_img = model.compute_mahalanobis_distances(feature_image)
+
+            # Segment out the areas of the image that fits well enough.
+            gui.thresh_val, segmented = perform_segmentation(mahalanobis_img, gui.thresh_val, use_otsu, max_distance)
+
+            # Highlight the segmented area in green in the input frame and draw the sampling rectangle.
+            frame[segmented > 0] *= np.uint8([0, 1, 0])
+            draw_sampling_rectangle(frame, sampling_rectangle)
+
+            # Normalise the Mahalanobis image to [0, max_distance] for visualisation.
+            mahalanobis_img = mahalanobis_img / max_distance
+
+            # Show the results
+            gui.show_frame(frame)
+            gui.show_mahalanobis(mahalanobis_img)
+
+            # Update the GUI and wait a short time for input from the keyboard.
+            key = gui.wait_key(1)
+
+            # React to keyboard commands.
+            if key == ord('q'):
+                print("Quitting")
+                break
+
+            elif key == ord(' '):
+                print("Extracting samples manually")
+                samples = extract_training_samples(feature_image, sampling_rectangle)
+                model = model_type(samples)
+
+            elif key == ord('o'):
+                use_otsu = not use_otsu
+                print(f"Use Otsu's: {use_otsu}")
+
+            elif key == ord('a'):
+                use_adaptive_model = not use_adaptive_model
+                print(f"Use adaptive model: {use_adaptive_model}")
+
+    # Stop video source.
+    cap.release()
+
+
+class MultivariateNormalModel:
+    """Represents a multivariate normal model"""
+
+    def __init__(self, samples):
+        """Constructs the model by training it on a set of feature samples
+
+        :param samples: A set of feature samples
+        """
+
+        self._perform_training(samples)
+
+    def _perform_training(self, samples):
+        """Trains the model"""
+
+        # TODO 1.1: Train the multivariate normal model by estimating the mean and covariance given the samples.
```

Comment:
Bedre med "... __for__ the given samples."?

Reply:
Nope ;)